### PR TITLE
[TypeScript] full support ✨🤖

### PR DESCRIPTION
**Pull request by Mimikkk**
I've been working for a couple of hours to add typescript support to the project.
It encompasses tests, utils and so on. Bumped vitest to 1.1.1 and did a few minor tweaks due to typing.

**Changes**:
- `typescript 5.3.3` support.
- `vite` for building minified solution in both commonjs and esmodules formats.
- `prettier` for code formatting.
- `concurrently/rimraf/copyfiles` for support building on linux/windows/macos.
- `vite` allowed me to split files into `utils` / `types` and `index` for ease of development

**Added scripts**:
- `build`: builds the package into the build directory using vite. It clears the `build/` directory, then it runs build process,
- `build:clear`: clears the `build/` directory
- `build:types`: emits the typescript definitions into the `build/` directory
- `build:lib`: build and minifies the package into the `build/` directory

I moved the dynamically created class `Chat` outside the `Gemini` class and added exported createGemini function for the ease of creation multiple instances.

### CODE REVIEW

**Comment by yoroshikun** [comment]
This is great 👍  Hope it gets merged soon.

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{
+  "compilerOptions": {
+    "baseUrl": ".",
+    "inlineSourceMap": true,
+    "inlineSources": true,
+    "module": "ESNext",
```
Comment (yoroshikun) [inline tsconfig.json]: TS somewhat complains about this `Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext` ts 5.3.3

**Comment by yoroshikun** [comment]
looks like you need to modify the fullCoverage.test.js file? at least changing the line 
`import Gemini from '../index';`
to
`import Gemini from '../src/index';`

allows it to run~

File: package.json [hunk 2]
```diff
@@ -23,7 +36,14 @@
   "type": "module",
   "devDependencies": {
     "@biomejs/biome": "1.4.1",
-    "@vitest/coverage-v8": "^1.0.4",
-    "vitest": "^1.0.4"
+    "@types/node": "^20.10.6",
```
Comment (yoroshikun) [inline package.json:39]: since the project is actually using bun, should we also include bun types?

**Comment by yoroshikun** [comment]
You may also want to remove the old `fullCoverage.test.js` in this PR

File: vite.config.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { defineConfig } from 'vite';
```
Comment (yoroshikun) [inline vite.config.ts]: Unfortunately if using bun the cache likes to be tested, can remove it with config.

```suggestion
import { defineConfig } from 'vitest/config';
```

```
{
...
test: {
  exclude: ['*/~/**']
}
}

File: tsconfig.json (hunk 1 above)
Comment (Mimikkk) [reply tsconfig.json]: We could do it, it changes resolution model from directory specification into file specification:
```ts
import {} from 'abc';
```
to:
```ts
import {} from 'abc.js';
```

File: vite.config.ts (hunk 3 above)
Comment (Mimikkk) [reply vite.config.ts]: Ok, great idea

Comment (Mimikkk) [reply vite.config.ts]: https://github.com/oven-sh/bun/issues/4145

When running vitest with bun it just crashes in a fire.
I excluded the test from the config. so it does not catch fire

File: tsconfig.json (hunk 1 above)
Comment (Mimikkk) [reply tsconfig.json]: Ok, i've changed it to use nodenext directly. Try it on your build to verify it.

**Comment by taciturnaxolotl** [comment]
hey @EvanZhouDev can you merge this? I would love to have full typescript support on the official package.

**Comment by taciturnaxolotl** [comment]
I might be dumb but how do you build and publish the package @Mimikkk? I've tried with npm publish and vite build but they don't seem to work.

**Comment by EvanZhouDev** [comment]
Hello @yoroshikun @kcoderhtml @Mimikkk, sorry for the super late response.

I know most likely you're not using this package anymore, but I thought I'd still update you on what's going on. I've been extremely busy for the past few months on work and other projects, so I haven't been able to catch up with `gemini-ai` recently. Furthermore, I have not decided to merge this branch because of a few refactors I've been doing, as well as the fact that it has seemed to stray a bit too far from the original.

However, seeing as Google has released Gemini 1.5 Flash today, I will be coming back to this project. Expect a relatively large update in the next few weeks, featuring types, updated streaming support, as well as better support by default for the new models.

As for this,
> I might be dumb but how do you build and publish the package @Mimikkk? I've tried with npm publish and vite build but they don't seem to work.

only the original publisher can update the package to NPM 😅

**Comment by taciturnaxolotl** [comment]
Thanks! Looking forward to trying the updated package! 

> only the original publisher can update the package to NPM 😅

I believe I was trying to rebuild the package and deploy to my NPM account, but I can't remember at this point what I was trying to do :)

**Comment by EvanZhouDev** [comment]
I have finished the initial rewrite in TypeScript, which is now available on the `v2` branch. It supports a bunch of improvements and some new features, which I will have to document 😅... it is still completely backward compatible, but I thought a ground-up rewrite deserves a major version bump.

Edit: PR now available at #9

**Comment by taciturnaxolotl** [comment]
Thank you!!! Looking forward to trying it out!

**Comment by EvanZhouDev** [comment]
Full TypeScript support, along with a bunch of great new features are now available on NPM. Will be closing this for now. Feel free to continue on this thread (or start a new issue) if anything needs to be fixed!

**Pull request closed without merging**